## **TELCO CUSTOMER CHURN PREDICTION**


### **Business Understanding**

##### **Problem Statement**
Customer retention is at the heart of most business models in their effort to increase their profit or revenue margin. Presently, most companies leverage machine learning to build classification models to perform churn analysis on their customers. The highly competitive nature of the telecommunications industry makes retaining customers extremely crucial. This project involves accessing and analyzing customer churn data from multiple sources, building a robust classification model, and helping a telecommunication company predict customer churn to improve retention strategies. The objective is to help a telecommunication company understand customer churn and its impact on profitability. 

##### **Goal and Objectives**

•    To understand the current customer churn rate.

•    To identify factors (such as demographics, usage patterns, etc.) that influence customer churn aiming to gain a deeper understanding of customer behavior

•    To build a predictive machine learning model to predict customer churn for a telecommunications company to forecast which customers are likely to churn

##### **Stakeholders**
•	Company Executives and Management

•	Data Science and Analytics Team

•	Customer Service and Support Teams

•	Marketing and Sales and Advertisement Teams:

•	Finance 

•	Legal and Compliance Team

##### **Key Metrics and Success Criteria**

•  Accuracy Requirement:

•	This model must achieve an accuracy score of at least 85% when evaluated on balanced data, ensuring a high proportion of correct predictions.
•  F1 Score Benchmark

•	Models should attain an F1 score greater than 0.80 (80%), indicating a strong balance between precision and recall, which is crucial for handling both false positives and false negatives effectively.

•  ROC Curve Standard:

•	An ROC curve with an area under the curve (AUC) of 80% is desired, demonstrating the model's ability to generalize well and maintain a good balance between sensitivity and specificity.

•  Baseline Models Requirement:

•	At least four different baseline models should be developed to serve as benchmarks. These could include logistic regression, decision trees, support vector machines, and k-nearest neighbors, providing a range of reference points for comparison.

•  Hyperparameter Tuning Condition:

Hyperparameter tuning will be conducted only on those baseline models that achieve an F1 score above the 0.80 threshold. This ensures that tuning efforts are concentrated on models that show initial promise and meet the performance criteria.


##### **Hypothesis**

•  Null Hypothesis (Ho): There is a significant difference in churn rates between customers who have received promotional offers and those who have not.

 •Alternate Hypothesis (Ha) : Customers who have not received any promotional offers in the past year have a higher churn rate.

     




##### **Analytical Questions**
I.	What are the key demographic and behavioral characteristics of customers who churn compared to those who stay, and how do these characteristics vary across different customer segments?

Insights derrived can include whether certain age groups or regions are more prone to churn, or if specific behaviors (like low usage or frequent complaints) correlate with higher churn rates. 

II.	What are the financial implications of churn, and how can retention strategies be optimized to maximize customer lifetime value (CLV) across different customer segments?

Insights include quantifying the cost of acquiring new customers versus retaining existing ones, identifying high CLV segments that need focused retention efforts

III. Which factors have the highest influence on customer churn, and how do they interact with each other?

Insights include identifying the most important factors like service quality, pricing competitiveness that contribute to churn, and how they interact with each other.

IV.	How do external factors such as economic conditions, competitor actions, and market trends influence customer churn rates?

Insights include understanding the impact of economic conditions, competitor actions, and market trends on customer churn rates.

V.	Is there a correlation between the customer’s payment method and churn rate?

Insights include understanding the correlation between the customer’s payment method and churn rate.

##### **Scope and Constraints**
Some constraints of this project include, computational resources, model complexity, time limitations, stakeholder expectations, and ethical and legal considerations.

##### **Additional Information**

This project is to be completed in 4 weeks 

















### **Data Understanding**

#### Importations

In [30]:
 
 #Data manipulation and analysis
import pandas as pd
import numpy as np
 
# Database connectivity
import pyodbc
 
# Database ORM (optional)
from sqlalchemy import create_engine
 
# Data visualization
import matplotlib.pyplot as plt
import seaborn as sns
 
# Machine learning 
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix
 


#### Data Loading

In [31]:
#Connecting to the first database
# Now the sql query to get the data is what what you see below.
# Define the connection string
server = 'dap-projects-database.database.windows.net'
database = 'dapDB'
username = 'LP2_project'
password = 'Stat$AndD@t@Rul3'
conn_str = f'DRIVER={{ODBC Driver 17 for SQL Server}};SERVER={server};DATABASE={database};UID={username};PWD={password}'

# Connect to the database
conn = pyodbc.connect(conn_str)

# Query the data
query = "SELECT * FROM dbo.LP2_Telco_churn_first_3000"
data1 = pd.read_sql(query, conn)

# Close the connection
conn.close()

# Display the data
print(data1.head(1))

C:\Users\USER\AppData\Local\Temp\ipykernel_12864\2806874801.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data1 = pd.read_sql(query, conn)


   customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG  Female          False     True       False       1   

   PhoneService MultipleLines InternetService OnlineSecurity  ...  \
0         False          None             DSL          False  ...   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0            False       False       False           False  Month-to-month   

  PaperlessBilling     PaymentMethod MonthlyCharges  TotalCharges  Churn  
0             True  Electronic check          29.85         29.85  False  

[1 rows x 21 columns]


In [32]:
#Connecting to the second database

# URL of the CSV file
data2= "https://raw.githubusercontent.com/Azubi-Africa/Career_Accelerator_LP2-Classifcation/main/LP2_Telco-churn-second-2000.csv"

# Load the dataframe from the URL
data2= pd.read_csv(data2)

# Display the first row to verify
print(data2.head(1))

   customerID gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  5600-PDUJF   Male              0      No         No       6          Yes   

  MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0            No             DSL             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0         Yes          No              No  Month-to-month              Yes   

             PaymentMethod MonthlyCharges  TotalCharges Churn  
0  Credit card (automatic)           49.5         312.7    No  

[1 rows x 21 columns]


In [33]:
#Connecting to the third database

# file path of the Excel file
file_path = r"C:\Users\USER\Desktop\Telco-churn-last-2000.xlsx"

# Load Excel file into a DataFrame
Test_data = pd.read_excel(file_path)

# Display the first few rows of the DataFrame
print(Test_data.head(1))


   customerID gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7613-LLQFO   Male              0      No         No      12          Yes   

  MultipleLines InternetService OnlineSecurity OnlineBackup DeviceProtection  \
0           Yes     Fiber optic             No           No               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No         Yes              No  Month-to-month              Yes   

      PaymentMethod  MonthlyCharges TotalCharges  
0  Electronic check           84.45      1059.55  


#### **Exploratory Data Analysis (EDA)**

- Data Quality Assement ,EDA & Data Cleaning

In [34]:
#Checking basic information for the fistdata set 
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        3000 non-null   object 
 1   gender            3000 non-null   object 
 2   SeniorCitizen     3000 non-null   bool   
 3   Partner           3000 non-null   bool   
 4   Dependents        3000 non-null   bool   
 5   tenure            3000 non-null   int64  
 6   PhoneService      3000 non-null   bool   
 7   MultipleLines     2731 non-null   object 
 8   InternetService   3000 non-null   object 
 9   OnlineSecurity    2349 non-null   object 
 10  OnlineBackup      2349 non-null   object 
 11  DeviceProtection  2349 non-null   object 
 12  TechSupport       2349 non-null   object 
 13  StreamingTV       2349 non-null   object 
 14  StreamingMovies   2349 non-null   object 
 15  Contract          3000 non-null   object 
 16  PaperlessBilling  3000 non-null   bool   


In [35]:
#Checking basic information for the second data set 
data2.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2043 entries, 0 to 2042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        2043 non-null   object 
 1   gender            2043 non-null   object 
 2   SeniorCitizen     2043 non-null   int64  
 3   Partner           2043 non-null   object 
 4   Dependents        2043 non-null   object 
 5   tenure            2043 non-null   int64  
 6   PhoneService      2043 non-null   object 
 7   MultipleLines     2043 non-null   object 
 8   InternetService   2043 non-null   object 
 9   OnlineSecurity    2043 non-null   object 
 10  OnlineBackup      2043 non-null   object 
 11  DeviceProtection  2043 non-null   object 
 12  TechSupport       2043 non-null   object 
 13  StreamingTV       2043 non-null   object 
 14  StreamingMovies   2043 non-null   object 
 15  Contract          2043 non-null   object 
 16  PaperlessBilling  2043 non-null   object 


In [36]:
#Checking basic information for the third data set 

Test_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        2000 non-null   object 
 1   gender            2000 non-null   object 
 2   SeniorCitizen     2000 non-null   int64  
 3   Partner           2000 non-null   object 
 4   Dependents        2000 non-null   object 
 5   tenure            2000 non-null   int64  
 6   PhoneService      2000 non-null   object 
 7   MultipleLines     2000 non-null   object 
 8   InternetService   2000 non-null   object 
 9   OnlineSecurity    2000 non-null   object 
 10  OnlineBackup      2000 non-null   object 
 11  DeviceProtection  2000 non-null   object 
 12  TechSupport       2000 non-null   object 
 13  StreamingTV       2000 non-null   object 
 14  StreamingMovies   2000 non-null   object 
 15  Contract          2000 non-null   object 
 16  PaperlessBilling  2000 non-null   object 


In [37]:
#concatenating data1 and data2

train_data = pd.concat([data1, data2], axis = 0)
train_data.tail()

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
2038  6840-RESVB    Male              0     Yes        Yes      24   
2039  2234-XADUH  Female              0     Yes        Yes      72   
2040  4801-JZAZL  Female              0     Yes        Yes      11   
2041  8361-LTMKD    Male              1     Yes         No       4   
2042  3186-AJIEK    Male              0      No         No      66   

     PhoneService     MultipleLines InternetService OnlineSecurity  ...  \
2038          Yes               Yes             DSL            Yes  ...   
2039          Yes               Yes     Fiber optic             No  ...   
2040           No  No phone service             DSL            Yes  ...   
2041          Yes               Yes     Fiber optic             No  ...   
2042          Yes                No     Fiber optic            Yes  ...   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
2038              Yes         Yes         Yes             Yes        One year   
2039              Yes          No         Yes             Yes        One year   
2040               No          No          No              No  Month-to-month   
2041               No          No          No              No  Month-to-month   
2042              Yes         Yes         Yes             Yes        Two year   

     PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
2038              Yes               Mailed check          84.80        1990.5   
2039              Yes    Credit card (automatic)         103.20        7362.9   
2040              Yes           Electronic check          29.60        346.45   
2041              Yes               Mailed check          74.40         306.6   
2042              Yes  Bank transfer (automatic)         105.65        6844.5   

     Churn  
2038    No  
2039    No  
2040    No  
2041   Yes  
2042    No  

[5 rows x 21 columns]

In [38]:
#Checking basic information for the train data set 

train_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5043 entries, 0 to 2042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        5043 non-null   object 
 1   gender            5043 non-null   object 
 2   SeniorCitizen     5043 non-null   int64  
 3   Partner           5043 non-null   object 
 4   Dependents        5043 non-null   object 
 5   tenure            5043 non-null   int64  
 6   PhoneService      5043 non-null   object 
 7   MultipleLines     4774 non-null   object 
 8   InternetService   5043 non-null   object 
 9   OnlineSecurity    4392 non-null   object 
 10  OnlineBackup      4392 non-null   object 
 11  DeviceProtection  4392 non-null   object 
 12  TechSupport       4392 non-null   object 
 13  StreamingTV       4392 non-null   object 
 14  StreamingMovies   4392 non-null   object 
 15  Contract          5043 non-null   object 
 16  PaperlessBilling  5043 non-null   object 
 17  

In [39]:
#viewing data columns of train dataset 

train_data.columns


Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [40]:
#Describe train data set
train_data.describe(include='all').T




count  unique               top  freq       mean  \
customerID          5043    5043        3186-AJIEK     1        NaN   
gender              5043       2              Male  2559        NaN   
SeniorCitizen     5043.0     NaN               NaN   NaN   0.162403   
Partner             5043       4             False  1538        NaN   
Dependents          5043       4             False  2070        NaN   
tenure            5043.0     NaN               NaN   NaN  32.576641   
PhoneService        5043       4              True  2731        NaN   
MultipleLines       4774       5             False  1437        NaN   
InternetService     5043       3       Fiber optic  2248        NaN   
OnlineSecurity      4392       5             False  1478        NaN   
OnlineBackup        4392       5             False  1320        NaN   
DeviceProtection    4392       5             False  1296        NaN   
TechSupport         4392       5             False  1476        NaN   
StreamingTV         4392       5             False  1190        NaN   
StreamingMovies     4392       5              True  1199        NaN   
Contract            5043       3    Month-to-month  2744        NaN   
PaperlessBilling    5043       4              True  1776        NaN   
PaymentMethod       5043       4  Electronic check  1700        NaN   
MonthlyCharges    5043.0     NaN               NaN   NaN  65.090214   
TotalCharges      5038.0  4884.0         20.200001   5.0        NaN   
Churn               5042       4             False  2219        NaN   

                        std   min     25%        50%        75%         max  
customerID              NaN   NaN     NaN        NaN        NaN         NaN  
gender                  NaN   NaN     NaN        NaN        NaN         NaN  
SeniorCitizen      0.368857   0.0     0.0        0.0        0.0         1.0  
Partner                 NaN   NaN     NaN        NaN        NaN         NaN  
Dependents              NaN   NaN     NaN        NaN        NaN         NaN  
tenure            24.529807   0.0     9.0       29.0       56.0        72.0  
PhoneService            NaN   NaN     NaN        NaN        NaN         NaN  
MultipleLines           NaN   NaN     NaN        NaN        NaN         NaN  
InternetService         NaN   NaN     NaN        NaN        NaN         NaN  
OnlineSecurity          NaN   NaN     NaN        NaN        NaN         NaN  
OnlineBackup            NaN   NaN     NaN        NaN        NaN         NaN  
DeviceProtection        NaN   NaN     NaN        NaN        NaN         NaN  
TechSupport             NaN   NaN     NaN        NaN        NaN         NaN  
StreamingTV             NaN   NaN     NaN        NaN        NaN         NaN  
StreamingMovies         NaN   NaN     NaN        NaN        NaN         NaN  
Contract                NaN   NaN     NaN        NaN        NaN         NaN  
PaperlessBilling        NaN   NaN     NaN        NaN        NaN         NaN  
PaymentMethod           NaN   NaN     NaN        NaN        NaN         NaN  
MonthlyCharges    30.068133  18.4  35.775  70.550003  90.050003  118.650002  
TotalCharges            NaN   NaN     NaN        NaN        NaN         NaN  
Churn                   NaN   NaN     NaN        NaN        NaN         NaN

In [41]:
#Checking for duplicates in train data set

train_data.duplicated().sum()

np.int64(0)

In [42]:
#Checking for missing values in train data set 

train_data.isnull().sum()

customerID            0
gender                0
SeniorCitizen         0
Partner               0
Dependents            0
tenure                0
PhoneService          0
MultipleLines       269
InternetService       0
OnlineSecurity      651
OnlineBackup        651
DeviceProtection    651
TechSupport         651
StreamingTV         651
StreamingMovies     651
Contract              0
PaperlessBilling      0
PaymentMethod         0
MonthlyCharges        0
TotalCharges          5
Churn                 1
dtype: int64

In [43]:
#Calculating the percentage of mising values in train data set 

train_data.isnull().sum()/len(train_data)*100

customerID           0.000000
gender               0.000000
SeniorCitizen        0.000000
Partner              0.000000
Dependents           0.000000
tenure               0.000000
PhoneService         0.000000
MultipleLines        5.334127
InternetService      0.000000
OnlineSecurity      12.908983
OnlineBackup        12.908983
DeviceProtection    12.908983
TechSupport         12.908983
StreamingTV         12.908983
StreamingMovies     12.908983
Contract             0.000000
PaperlessBilling     0.000000
PaymentMethod        0.000000
MonthlyCharges       0.000000
TotalCharges         0.099147
Churn                0.019829
dtype: float64

In [44]:
def standardize_value(value):
    if value is True or value == 'Yes':
        return 'Yes'
    elif value is False or value == 'No' or value == 'None':
        return 'No'
    elif value == 'No internet service' or value == 'No phone service':
        return value 
    else:
        return 'Unknown'

# List of columns to be standardized
columns_to_standardize = [
    'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 
    'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 
    'StreamingTV', 'StreamingMovies', 'PaperlessBilling', 'Churn'
]

# Apply standardization to each column in the list
for column in columns_to_standardize:
    train_data[column] = train_data[column].apply(standardize_value)

# Confirm changes in merged data set
for column in columns_to_standardize:
    unique_values_after = train_data[column].unique()
    print(f'Unique values in {column} after standardization: {unique_values_after}')
    print (train_data[column].value_counts())
    print()

Unique values in Partner after standardization: ['Yes' 'No']
Partner
No     2585
Yes    2458
Name: count, dtype: int64

Unique values in Dependents after standardization: ['No' 'Yes']
Dependents
No     3482
Yes    1561
Name: count, dtype: int64

Unique values in PhoneService after standardization: ['No' 'Yes']
PhoneService
Yes    4554
No      489
Name: count, dtype: int64

Unique values in MultipleLines after standardization: ['Unknown' 'No' 'Yes' 'No phone service']
MultipleLines
No                  2424
Yes                 2130
Unknown              269
No phone service     220
Name: count, dtype: int64

Unique values in OnlineSecurity after standardization: ['No' 'Yes' 'Unknown' 'No internet service']
OnlineSecurity
No                     2507
Yes                    1456
Unknown                 651
No internet service     429
Name: count, dtype: int64

Unique values in OnlineBackup after standardization: ['Yes' 'No' 'Unknown' 'No internet service']
OnlineBackup
No                    

In [45]:
#Change total charges to numeric i.e. form obj to float

train_data['TotalCharges'] = pd.to_numeric(train_data['TotalCharges'], errors='coerce')
train_data['TotalCharges'] = train_data['TotalCharges'].astype(float)
train_data.head(1)

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   

  MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0       Unknown             DSL             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   

      PaymentMethod MonthlyCharges  TotalCharges  Churn  
0  Electronic check          29.85         29.85     No  

[1 rows x 21 columns]

In [46]:
#checking for unique values in train data set

train_data.nunique()

customerID          5043
gender                 2
SeniorCitizen          2
Partner                2
Dependents             2
tenure                73
PhoneService           2
MultipleLines          4
InternetService        3
OnlineSecurity         4
OnlineBackup           4
DeviceProtection       4
TechSupport            4
StreamingTV            4
StreamingMovies        4
Contract               3
PaperlessBilling       2
PaymentMethod          4
MonthlyCharges      2069
TotalCharges        4863
Churn                  3
dtype: int64

In [47]:
#Check for null values in train data set
train_data.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        8
Churn               0
dtype: int64

In [48]:
#Filling missing values in merged data set (Total Charges)

train_data['TotalCharges'] = train_data['TotalCharges'].fillna(train_data['TotalCharges'].mean())

#Check for null values in train data set

train_data.isnull().sum()


customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [49]:
#descbribe train data set only showing the numerical variables

train_data.describe().T



count         mean          std        min      25%  \
SeniorCitizen   5043.0     0.162403     0.368857   0.000000    0.000   
tenure          5043.0    32.576641    24.529807   0.000000    9.000   
MonthlyCharges  5043.0    65.090214    30.068133  18.400000   35.775   
TotalCharges    5043.0  2302.062861  2267.682399  18.799999  418.000   

                        50%          75%          max  
SeniorCitizen      0.000000     0.000000     1.000000  
tenure            29.000000    56.000000    72.000000  
MonthlyCharges    70.550003    90.050003   118.650002  
TotalCharges    1406.000000  3857.575073  8670.100000

In [50]:
#describe train data set only showing the categorical variables
train_data.describe(include='object').T

count unique               top  freq
customerID        5043   5043        3186-AJIEK     1
gender            5043      2              Male  2559
Partner           5043      2                No  2585
Dependents        5043      2                No  3482
PhoneService      5043      2               Yes  4554
MultipleLines     5043      4                No  2424
InternetService   5043      3       Fiber optic  2248
OnlineSecurity    5043      4                No  2507
OnlineBackup      5043      4                No  2231
DeviceProtection  5043      4                No  2219
TechSupport       5043      4                No  2472
StreamingTV       5043      4                No  2018
StreamingMovies   5043      4               Yes  1982
Contract          5043      3    Month-to-month  2744
PaperlessBilling  5043      2               Yes  2996
PaymentMethod     5043      4  Electronic check  1700
Churn             5043      3                No  3706

#### Key Insights and Observations
- The data for this telco customer prediction analysis are 3 sets of data derrived from different sources. The first dataset (data1) had 21 columns and 3000 row and the second dataset (data2) had 21 columns and 2043 rows. The last set is known as Test data would be used later in module build. The first 2 datasets were merged to create a new dataset (train_data) with 21 columns and 5043 rows

- Categorical values from the train_data are: customerID, gender, Partner, Dependents, PhoneService, MultipleLines, InternetService, OnlineSecurity, OnlineBackup, DeviceProtection, TechSupport, StreamingTV, StreamingMovies, Contract, PaperlessBilling, PaymentMethod, Churn

- Numerical values from the train_data are: SeniorCitizen, tenure, MonthlyCharges, TotalCharges

- After merger it was noticed these columns had null values: MultipleLines: 269 null values, OnlineSecurity: 651 null values, OnlineBackup: 651 null values, DeviceProtection: 651 null values, TechSupport: 651 null values, StreamingTV: 651 null values, StreamingMovies: 651, null values, TotalCharges: 5 null values, Churn: 1 null value. 

- Columns were standardized in above function which helped with missing values in all columns except TotalCharges column..

- The TotalCharges column had 8 null values which were filled with the mean of the column and column was changed from object type to float type.

- The SeniorCitizen column indicated binary variables indicationg 1 as senior citizen or not as 0. The mean value of approximately 0.162 indicates that 16.2% of the customers are senior citizens.. The standard deviation of 0.354 indicates that the data is not normally distributed. The quatiles (25%, 50% and 75%) are 0 which shows most of the telco'c customers are not senior citizens.

- The tenure column indicated the number of months the customer has been with the company. The mean value of approximately 32.5 months indicates that the average tenure is 29.8 months. The standard deviation of 24.5 months indicates that the data is not normally distributed. The median value of 29.8 months indicates that 50% of the customers have been with the company for less than 29 months. Minimum tenure is 0 month and maximum tenure is 72 months.

- The MonthlyCharges column indicated the amount charged to the customer monthly. The mean value of approximately 65.09 indicates with a standard deviation of 30.01 indicates again, that the data is not normally distributed. The median monthly charge is $70.35. Minimum monthly charge is $18.4. The range spans from a minimum of $18.4 to a maximum of $118.8. Thisindicates that the data is skewed to the right.

- The TotalCharges column indicated the total amount charged to the customer. The mean value of approximately 2302.06 indicates with a standard deviation of 2267.6 indicates again, that the data is not normally distributed. The median total charge is 1406.0 Minimum total charge is $18.80. This range of values indicates that the data is skewed to the right.